Implement CanonicalSerialize/Deserialize improvements

paberr opened the pull request:
Draft for #104.
Not ready yet, I still want to add some tests for uncompressed serialization/deserialization.

Feedback is already welcome.

TODOs:
- [x] Rebase to current master.
- [x] Tests for uncompressed serialization/deserialization.
- [x] Improve Fp serialization (do not serialize full u64 limbs).
- [x] Simple derive macro.

paberr committed:
Implement CanonicalSerialize/Deserialize improvements

Pratyush commented on algebra/src/curves/models/twisted_edwards_extended/tests.rs:
```diff
@@ -33,27 +28,25 @@ pub fn edwards_curve_serialization_test<P: TEModelParameters>(buf_size: usize) {
         {
             let mut serialized = vec![0; buf_size];
-            a.serialize(&[], &mut serialized).unwrap();
+            a.serialize(&mut &mut serialized[..]).unwrap();
```
```suggestion
            a.serialize(&mut serialized[..]).unwrap();
```

paberr replied:
FYI: the second `&mut` is required here as it needs to get a mutable pointer to a mutable slice here.
However, I agree that it is very hard to read. For better readability, I replaced it by using a cursor.
```
let mut cursor = Cursor::new(&mut serialized[..]);
a.serialize(&mut cursor).unwrap();
```

Pratyush commented on algebra/src/curves/models/twisted_edwards_extended/tests.rs:
```diff
@@ -33,27 +28,25 @@ pub fn edwards_curve_serialization_test<P: TEModelParameters>(buf_size: usize) {
 
-            let mut extra_info_buf = [false; 0];
-            let b = GroupAffine::<P>::deserialize(&serialized, &mut extra_info_buf).unwrap();
+            let b = GroupAffine::<P>::deserialize(&mut &serialized[..]).unwrap();
```
```suggestion
            let b = GroupAffine::<P>::deserialize(&mut serialized[..]).unwrap();
```

Pratyush commented on algebra/src/curves/models/twisted_edwards_extended/tests.rs:
```diff
@@ -33,27 +28,25 @@ pub fn edwards_curve_serialization_test<P: TEModelParameters>(buf_size: usize) {
-            let mut extra_info_buf = [false; 0];
-            let b = GroupAffine::<P>::deserialize(&serialized, &mut extra_info_buf).unwrap();
+            a.serialize(&mut &mut serialized[..]).unwrap();
+            let b = GroupAffine::<P>::deserialize(&mut &serialized[..]).unwrap();
```
```suggestion
            let b = GroupAffine::<P>::deserialize(&mut serialized[..]).unwrap();
```

Pratyush commented on algebra/src/curves/models/twisted_edwards_extended/tests.rs:
```diff
@@ -33,27 +28,25 @@ pub fn edwards_curve_serialization_test<P: TEModelParameters>(buf_size: usize) {
-            assert!(if let SerializationError::ExtraInfoWrongSize =
-                a.serialize(&[false; 1], &mut serialized).unwrap_err()
+            assert!(if let SerializationError::UnexpectedFlags =
+                a.serialize_with_flags(&mut &mut serialized[..], EdwardsFlags::default()).unwrap_err()
```
```suggestion
                a.serialize_with_flags(&mut serialized[..], EdwardsFlags::default()).unwrap_err()
```

Pratyush commented on algebra/src/curves/tests.rs:
```diff
@@ -294,36 +289,33 @@ pub fn sw_curve_serialization_test<P: SWModelParameters>(buf_size: usize) {
         {
             let mut serialized = vec![0; buf_size];
-            a.serialize(&[], &mut serialized).unwrap();
+            a.serialize(&mut &mut serialized[..]).unwrap();
```
```suggestion
            a.serialize(&mut serialized[..]).unwrap();
```

Pratyush commented on algebra/src/curves/tests.rs:
```diff
@@ -294,36 +289,33 @@ pub fn sw_curve_serialization_test<P: SWModelParameters>(buf_size: usize) {
 
-            let mut extra_info_buf = [false; 0];
-            let b = GroupAffine::<P>::deserialize(&serialized, &mut extra_info_buf).unwrap();
+            let b = GroupAffine::<P>::deserialize(&mut &serialized[..]).unwrap();
```
```suggestion
            let b = GroupAffine::<P>::deserialize(&mut serialized[..]).unwrap();
```

Pratyush commented on algebra/src/curves/tests.rs:
```diff
@@ -294,36 +289,33 @@ pub fn sw_curve_serialization_test<P: SWModelParameters>(buf_size: usize) {
-            let mut extra_info_buf = [false; 0];
-            let b = GroupAffine::<P>::deserialize(&serialized, &mut extra_info_buf).unwrap();
+            a.serialize(&mut &mut serialized[..]).unwrap();
+            let b = GroupAffine::<P>::deserialize(&mut &serialized[..]).unwrap();
```
```suggestion
            let b = GroupAffine::<P>::deserialize(&mut serialized[..]).unwrap();
```

Pratyush commented on algebra/src/curves/tests.rs:
```diff
@@ -294,36 +289,33 @@ pub fn sw_curve_serialization_test<P: SWModelParameters>(buf_size: usize) {
-            let mut extra_info_buf = [false; 0];
-            let b = GroupAffine::<P>::deserialize(&serialized, &mut extra_info_buf).unwrap();
+            a.serialize(&mut &mut serialized[..]).unwrap();
+            let b = GroupAffine::<P>::deserialize(&mut &serialized[..]).unwrap();
```
```suggestion
            let b = GroupAffine::<P>::deserialize(&mut serialized[..]).unwrap();
```

Pratyush commented on algebra/src/curves/tests.rs:
```diff
@@ -294,36 +289,33 @@ pub fn sw_curve_serialization_test<P: SWModelParameters>(buf_size: usize) {
-            assert!(if let SerializationError::ExtraInfoWrongSize =
-                a.serialize(&[false; 1], &mut serialized).unwrap_err()
+            assert!(if let SerializationError::UnexpectedFlags =
+                a.serialize_with_flags(&mut &mut serialized[..], SWFlags::default()).unwrap_err()
```
```suggestion
                a.serialize_with_flags(&mut serialized[..], SWFlags::default()).unwrap_err()
```

Pratyush commented on algebra/src/curves/tests.rs:
```diff
@@ -333,40 +325,13 @@ pub fn sw_curve_serialization_test<P: SWModelParameters>(buf_size: usize) {
-                false
-            });
+            let mut serialized = vec![0; buf_size - 1];
+            a.serialize(&mut &mut serialized[..]).unwrap_err();
```
```suggestion
            a.serialize(&mut serialized[..]).unwrap_err();
```

Pratyush commented on algebra/src/curves/tests.rs:
```diff
@@ -333,40 +325,13 @@ pub fn sw_curve_serialization_test<P: SWModelParameters>(buf_size: usize) {
-                false
-            });
+            let serialized = vec![0; buf_size - 1];
+            GroupAffine::<P>::deserialize(&mut &serialized[..]).unwrap_err();
```
```suggestion
            GroupAffine::<P>::deserialize(&mut serialized[..]).unwrap_err();
```

Pratyush commented on algebra/src/fields/tests.rs:
```diff
@@ -360,36 +361,54 @@ pub fn field_serialization_test<F: Field>(buf_size: usize) {
-            let mut serialized = vec![0; buf_size];
-            a.serialize(&[], &mut serialized).unwrap();
+            let mut serialized = vec![0u8; buf_size];
+            a.serialize(&mut &mut serialized[..]).unwrap();
```
```suggestion
            a.serialize(&mut serialized[..]).unwrap();
```

Pratyush commented on algebra/src/fields/tests.rs:
```diff
@@ -360,36 +361,54 @@ pub fn field_serialization_test<F: Field>(buf_size: usize) {
 
-            let mut extra_info_buf = [false; 0];
-            let b = F::deserialize(&serialized, &mut extra_info_buf).unwrap();
+            let b = F::deserialize(&mut &serialized[..]).unwrap();
```
```suggestion
            let b = F::deserialize(&mut serialized[..]).unwrap();
```

Pratyush commented on algebra/src/fields/tests.rs:
```diff
@@ -360,36 +361,54 @@ pub fn field_serialization_test<F: Field>(buf_size: usize) {
-            assert_eq!(extra_info_buf, [true; 2]);
+            let mut serialized = vec![0u8; buf_size];
+            a.serialize_with_flags(&mut &mut serialized[..], SWFlags::y_sign(true)).unwrap();
+            let (b, flags) = F::deserialize_with_flags::<_, SWFlags>(&mut &serialized[..]).unwrap();
```
```suggestion
            let (b, flags) = F::deserialize_with_flags::<_, SWFlags>(&mut serialized[..]).unwrap();
```

Pratyush commented on algebra/src/fields/tests.rs:
```diff
@@ -360,36 +361,54 @@ pub fn field_serialization_test<F: Field>(buf_size: usize) {
             let mut serialized = vec![0; buf_size];
             assert!(if let SerializationError::NotEnoughSpace =
-                a.serialize(&[true; 200], &mut serialized).unwrap_err()
+                a.serialize_with_flags(&mut &mut serialized[..], DummyFlags).unwrap_err()
```
```suggestion
                a.serialize_with_flags(&mut serialized[..], DummyFlags).unwrap_err()
```

Pratyush commented on algebra/src/fields/tests.rs:
```diff
@@ -360,36 +361,54 @@ pub fn field_serialization_test<F: Field>(buf_size: usize) {
-            let mut extra_info_buf = [false; 200];
             assert!(if let SerializationError::NotEnoughSpace =
-                F::deserialize(&serialized, &mut extra_info_buf).unwrap_err()
+                F::deserialize_with_flags::<_, DummyFlags>(&mut &serialized[..]).unwrap_err()
```
```suggestion
                F::deserialize_with_flags::<_, DummyFlags>(&mut serialized[..]).unwrap_err()
```

Pratyush commented on algebra/src/fields/tests.rs:
```diff
@@ -398,23 +417,10 @@ pub fn field_serialization_test<F: Field>(buf_size: usize) {
-                false
-            });
+            let mut serialized = vec![0; buf_size - 1];
+            a.serialize(&mut &mut serialized[..]).unwrap_err();
```
```suggestion
            a.serialize(&mut serialized[..]).unwrap_err();
```

Pratyush commented on algebra/src/fields/tests.rs:
```diff
@@ -398,23 +417,10 @@ pub fn field_serialization_test<F: Field>(buf_size: usize) {
-            } else {
-                false
-            });
+            F::deserialize(&mut &serialized[..]).unwrap_err();
```
```suggestion
            F::deserialize(&mut serialized[..]).unwrap_err();
```

Pratyush reviewed (changes requested):
This is great, thanks! I left some nits. Could you also rebase? #100 split up algebra, so I'd like to review the rebased PR to make sure nothing goes wrong during an automatic rebase.

kobigurk commented on algebra/src/biginteger/mod.rs:
```diff
@@ -21,11 +16,52 @@ bigint_impl!(BigInteger384, 6);
 bigint_impl!(BigInteger768, 12);
 bigint_impl!(BigInteger832, 13);
 
+impl<T: BigInteger> CanonicalSerialize for T {
```
Great PR!

Does this replace the Fp serialization?

If so, there's a significant disadvantage to just relying on the limbs for the serialized size. BigIntegers are often longer than the requires bytes needed to encode the field elements. Since big integers sizes are multiples of 64 and the field size is a multiple of 8, there would be always-0 bytes in the last limb, which affects the serialized size significantly.

Pratyush replied:
Did the old implementation avoid this pitfall?

paberr replied:
As far as I can see, the old Fp serialization simply called `self.write(&mut bytes[..])?;` (from the `ToBytes` trait), which in turn calls `ToBytes` on the `.into_repr()` (which is a BigInteger).
So, it probably has the same pitfall right now. But it is an interesting point!

kobigurk replied:
We avoided it in the past because we require the output_buf to have length equal to the byte size of the field, and copied from the 64-bit aligned serialized struct:

https://github.com/scipr-lab/zexe/blob/603637946a8e88685df75a0d7b007751730bc286/algebra-core/src/serialize/mod.rs#L39

paberr replied:
Ok, good catch then. I’ll add this aspect to my serialization as well.

paberr committed:
Improve code readability by using cursor

paberr committed:
Improve formatting

paberr committed:
Add tests for uncompressed serialization

Pratyush commented on algebra-core/src/io.rs:
```diff
@@ -101,3 +101,21 @@ impl Write for Cursor<&mut [u8]> {
+}
+
+impl Read for Cursor<&[u8]> {
+    fn read_exact(&mut self, data: &mut [u8]) -> Result<()> {
```
I believe the PR is failing because of right now?

```suggestion
    fn read_exact(&mut self, buf: &mut [u8]) -> Result<()> {
```

Pratyush commented:
Oh and can you delete `algebra/src/curves`? It's an artifact from the rebase, I think?

Pratyush commented on algebra-core/src/serialize/mod.rs:
```diff
@@ -24,70 +117,48 @@ pub fn buffer_bit_byte_size(modulus_bits: usize) -> (usize, usize) {
-
-                output_buf.copy_from_slice(&bytes[..output_byte_size]);
-                Ok(())
+                CanonicalSerialize::serialize_with_flags(&self.into_repr(), writer, flags)
```
```suggestion
                self.into_repr().serialize_with_flags(writer, flags)
```

Pratyush commented on algebra-core/src/serialize/mod.rs:
```diff
@@ -24,70 +117,48 @@ pub fn buffer_bit_byte_size(modulus_bits: usize) -> (usize, usize) {
-                    crate::serialize::buffer_bit_byte_size($field::<P>::size_in_bits());
-                byte_size
+            fn serialized_size(&self) -> usize {
+                CanonicalSerialize::serialized_size(&self.into_repr())
```
```suggestion
                self.into_repr().serialized_size()
```

Pratyush commented on algebra-core/src/fields/models/fp12_2over3over2.rs:
```diff
@@ -487,41 +487,34 @@ impl<P: Fp12Parameters> FromBytes for Fp12<P> {
-            &mut output_buf[6 * fp_byte_size..12 * fp_byte_size],
-        )?;
+        CanonicalSerialize::serialize(&self.c0, writer)?;
+        CanonicalSerialize::serialize_with_flags(&self.c1, writer, flags)?;
```
I believe this should be fine right? The only reason to not use this form is if we expect multiple traits with this method to be in scope, which I don't think should be the case.
```suggestion
        self.c0.serialize(writer)?;
        self.c1.serialize_with_flags(writer, flags)?;
```

paberr replied:
Yes, both should be equivalent. I can change it to the other style.

Pratyush commented on algebra-core/src/fields/models/fp12_2over3over2.rs:
```diff
@@ -487,41 +487,34 @@ impl<P: Fp12Parameters> FromBytes for Fp12<P> {
-        12*<<<P::Fp6Params as Fp6Parameters>::Fp2Params as Fp2Parameters>::Fp as CanonicalSerialize>::buffer_size()
+    fn serialized_size(&self) -> usize {
+        CanonicalSerialize::serialized_size(&self.c0)
+            + CanonicalSerialize::serialized_size(&self.c1)
```
```suggestion
       self.c0.serialized_size() + self.c1.serialized_size()
```

Pratyush commented on algebra-core/src/fields/models/fp12_2over3over2.rs:
```diff
@@ -487,41 +487,34 @@ impl<P: Fp12Parameters> FromBytes for Fp12<P> {
-        let c1 = Fp6::deserialize(&bytes[6 * fp_byte_size..12 * fp_byte_size], extra_info_buf)?;
+    fn deserialize<R: Read>(reader: &mut R) -> Result<Self, SerializationError> {
+        let c0: Fp6<_> = CanonicalDeserialize::deserialize(reader)?;
+        let c1: Fp6<_> = CanonicalDeserialize::deserialize(reader)?;
```
```suggestion
        let c0 = Fp6::deserialize(reader)?;
        let c1 = Fp6::deserialize(reader)?;
```

Pratyush commented on algebra-core/src/fields/models/fp12_2over3over2.rs:
```diff
@@ -487,41 +487,34 @@ impl<P: Fp12Parameters> FromBytes for Fp12<P> {
+        reader: &mut R,
+    ) -> Result<(Self, F), SerializationError> {
+        let c0: Fp6<_> = CanonicalDeserialize::deserialize(reader)?;
+        let (c1, flags): (Fp6<_>, _) = CanonicalDeserialize::deserialize_with_flags(reader)?;
```
```suggestion
        let c0 = Fp6::deserialize(reader)?;
        let (c1, flags) = Fp6::deserialize_with_flags(reader)?;
```

Pratyush commented on algebra-core/src/fields/models/fp3.rs:
```diff
@@ -492,46 +494,38 @@ impl<P: Fp3Parameters> fmt::Display for Fp3<P> {
-        )?;
+        CanonicalSerialize::serialize(&self.c0, writer)?;
+        CanonicalSerialize::serialize(&self.c1, writer)?;
+        CanonicalSerialize::serialize_with_flags(&self.c2, writer, flags)?;
```
```suggestion
        self.c0.serialize(writer)?;
        self.c1.serialize(writer)?;
        self.c2.serialize_with_flags(writer, flags)?;
```

Pratyush commented on algebra-core/src/fields/models/fp3.rs:
```diff
@@ -492,46 +494,38 @@ impl<P: Fp3Parameters> fmt::Display for Fp3<P> {
+    fn serialized_size(&self) -> usize {
+        CanonicalSerialize::serialized_size(&self.c0)
+            + CanonicalSerialize::serialized_size(&self.c1)
+            + CanonicalSerialize::serialized_size(&self.c2)
```
```suggestion
        self.c0.serialized_size()
            + self.c1.serialized_size()
            + self.c2.serialized_size()
```

Pratyush commented on algebra-core/src/fields/models/fp6_2over3.rs:
```diff
@@ -400,43 +400,34 @@ impl<P: Fp6Parameters> fmt::Display for Fp6<P> {
-            &mut output_buf[3 * fp_byte_size..6 * fp_byte_size],
-        )?;
+        CanonicalSerialize::serialize(&self.c0, writer)?;
+        CanonicalSerialize::serialize_with_flags(&self.c1, writer, flags)?;
```
```suggestion
        self.c0.serialize(writer)?;
        self.c1.serialize_with_flags(writer, flags)?;
```

Pratyush commented on algebra-core/src/fields/models/fp6_2over3.rs:
```diff
@@ -400,43 +400,34 @@ impl<P: Fp6Parameters> fmt::Display for Fp6<P> {
-        6 * <<P::Fp3Params as Fp3Parameters>::Fp as CanonicalSerialize>::buffer_size()
+    fn serialized_size(&self) -> usize {
+        CanonicalSerialize::serialized_size(&self.c0)
+            + CanonicalSerialize::serialized_size(&self.c1)
```
```suggestion
        self.c0.serialized_size() + self.c1.serialized_size()
```

Pratyush commented on algebra-core/src/fields/models/fp6_2over3.rs:
```diff
@@ -400,43 +400,34 @@ impl<P: Fp6Parameters> fmt::Display for Fp6<P> {
-        let c1 = Fp3::deserialize(&bytes[3 * fp_byte_size..6 * fp_byte_size], extra_info_buf)?;
+    fn deserialize<R: Read>(reader: &mut R) -> Result<Self, SerializationError> {
+        let c0: Fp3<_> = CanonicalDeserialize::deserialize(reader)?;
+        let c1: Fp3<_> = CanonicalDeserialize::deserialize(reader)?;
```
```suggestion
        let c0 = Fp3::deserialize(reader)?;
        let c1 = Fp3::deserialize(reader)?;
```

Pratyush commented on algebra-core/src/fields/models/fp6_2over3.rs:
```diff
@@ -400,43 +400,34 @@ impl<P: Fp6Parameters> fmt::Display for Fp6<P> {
+        reader: &mut R,
+    ) -> Result<(Self, F), SerializationError> {
+        let c0: Fp3<_> = CanonicalDeserialize::deserialize(reader)?;
+        let (c1, flags): (Fp3<_>, _) = CanonicalDeserialize::deserialize_with_flags(reader)?;
```
```suggestion
        let c0 = Fp3::deserialize(reader)?;
        let (c1, flags) = Fp3::deserialize_with_flags(reader)?;
```

Pratyush commented on algebra-core/src/fields/models/fp6_3over2.rs:
```diff
@@ -477,49 +477,38 @@ impl<P: Fp6Parameters> FromBytes for Fp6<P> {
-        )?;
+        CanonicalSerialize::serialize(&self.c0, writer)?;
+        CanonicalSerialize::serialize(&self.c1, writer)?;
+        CanonicalSerialize::serialize_with_flags(&self.c2, writer, flags)?;
```
```suggestion
        self.c0.serialize(writer)?;
        self.c1.serialize(writer)?;
        self.c2.serialize_with_flags(writer, flags)?;
```

Pratyush commented on algebra-core/src/fields/models/fp6_3over2.rs:
```diff
@@ -477,49 +477,38 @@ impl<P: Fp6Parameters> FromBytes for Fp6<P> {
+    fn serialized_size(&self) -> usize {
+        CanonicalSerialize::serialized_size(&self.c0)
+            + CanonicalSerialize::serialized_size(&self.c1)
+            + CanonicalSerialize::serialized_size(&self.c2)
```
```suggestion
        self.c0.serialized_size()
            + self.c1.serialized_size()
            + self.c2.serialized_size()
```

Pratyush commented on algebra-core/src/fields/models/fp6_3over2.rs:
```diff
@@ -477,49 +477,38 @@ impl<P: Fp6Parameters> FromBytes for Fp6<P> {
+    fn deserialize<R: Read>(reader: &mut R) -> Result<Self, SerializationError> {
+        let c0: Fp2<_> = CanonicalDeserialize::deserialize(reader)?;
+        let c1: Fp2<_> = CanonicalDeserialize::deserialize(reader)?;
+        let c2: Fp2<_> = CanonicalDeserialize::deserialize(reader)?;
```
```suggestion
        let c0 = Fp2::deserialize(reader)?;
        let c1 = Fp2::deserialize(reader)?;
        let c2 = Fp2::deserialize(reader)?;
```

Pratyush commented on algebra-core/src/fields/models/fp6_3over2.rs:
```diff
@@ -477,49 +477,38 @@ impl<P: Fp6Parameters> FromBytes for Fp6<P> {
+    ) -> Result<(Self, F), SerializationError> {
+        let c0: Fp2<_> = CanonicalDeserialize::deserialize(reader)?;
+        let c1: Fp2<_> = CanonicalDeserialize::deserialize(reader)?;
+        let (c2, flags): (Fp2<_>, _) = CanonicalDeserialize::deserialize_with_flags(reader)?;
```
```suggestion
        let c0 = Fp2::deserialize(reader)?;
        let c1 = Fp2::deserialize(reader)?;
        let (c2, flags) = Fp2::deserialize_with_flags(reader)?;
```

Pratyush commented on algebra-core/src/serialize/flags.rs:
```diff
@@ -0,0 +1,125 @@
+pub struct SWFlags {
+    pub y_sign:      bool,
+    pub is_infinity: bool,
+}
```
Since the two cases are mutually exclusize, can we make this an enum?
```suggestion
pub enum SWFlags {
    Infinity,
    PositiveY,
    NegativeY,
}
```

Pratyush commented on algebra-core/src/serialize/flags.rs:
```diff
@@ -0,0 +1,125 @@
+#[derive(Default, Clone, Copy)]
+pub struct EdwardsFlags {
+    pub y_sign: bool,
+}
```
Similar Idea here:
```suggestion
pub enum EdwardsFlags {
    PositiveY,
    NegativeY,
}
```

Pratyush commented on algebra-core/src/serialize/mod.rs:
```diff
@@ -1,19 +1,112 @@
+
+    /// Reads `Self` from `reader` without compression.
+    fn deserialize_uncompressed<R: Read>(reader: &mut R) -> Result<Self, SerializationError> {
+        CanonicalDeserialize::deserialize(reader)
```
```suggestion
        Self::deserialize(reader)
```

Pratyush commented on algebra-core/src/serialize/mod.rs:
```diff
@@ -1,19 +1,112 @@
+
+    /// Serializes `self` into `writer` without compression.
+    fn serialize_uncompressed<W: Write>(&self, writer: &mut W) -> Result<(), SerializationError> {
+        CanonicalSerialize::serialize(self, writer)
```
```suggestion
        self.serialize(writer)
```

Pratyush commented on algebra-core/src/serialize/mod.rs:
```diff
@@ -1,19 +1,112 @@
-    fn serialize(
+    /// Serializes `self` into `writer`.
+    fn serialize<W: Write>(&self, writer: &mut W) -> Result<(), SerializationError> {
+        CanonicalSerialize::serialize_with_flags(self, writer, EmptyFlags::default())
```
```suggestion
        self.serialize_with_flags(writer, EmptyFlags::default())
```

Pratyush commented on algebra-core/src/serialize/mod.rs:
```diff
@@ -96,48 +167,86 @@ macro_rules! impl_prime_field_serializer {
-                        self.x.serialize(&[false, false], output_buf)
-                    }
+                    let flags = crate::serialize::SWFlags::y_sign(self.y > -self.y);
+                    CanonicalSerialize::serialize_with_flags(&self.x, writer, flags)
```
```suggestion
                    self.x.serialize_with_flags(writer, flags)
```

Pratyush commented on algebra-core/src/serialize/mod.rs:
```diff
@@ -96,48 +167,86 @@ macro_rules! impl_prime_field_serializer {
-            fn buffer_size() -> usize {
-                <P::BaseField as CanonicalSerialize>::buffer_size()
+            fn serialized_size(&self) -> usize {
+                CanonicalSerialize::serialized_size(&self.x)
```
```suggestion
                self.x.serialized_size()
```

Pratyush commented on algebra-core/src/serialize/mod.rs:
```diff
@@ -96,48 +167,86 @@ macro_rules! impl_prime_field_serializer {
+            ) -> Result<(), crate::serialize::SerializationError> {
+                CanonicalSerialize::serialize_uncompressed(&self.x, writer)?;
+                CanonicalSerialize::serialize_uncompressed(&self.y, writer)?;
+                CanonicalSerialize::serialize_uncompressed(&self.infinity, writer)?;
```
I think for this we still want to serialize to 2 field elements. That is, we want something like this:
```rust
if self.is_zero() {
    self.x.serialize_with_flags(writer, SWFlag::Zero)?;
    self.y.serialize(writer)?;
} else {
    self.x.serialize(writer)?;
    self.y.serialize(writer)?;
}
```

paberr replied:
True. We probably want to put the flags into the `y` though to be consistent with other serializations (the flags are usually encoded into the last element.

Pratyush commented on algebra/src/edwards_sw6/curves/tests.rs:
```diff
@@ -12,7 +12,7 @@ use crate::tests::{curves::*, groups::*};
     curve_tests::<EdwardsProjective>();
 
-    let byte_size = <EdwardsAffine as CanonicalSerialize>::buffer_size();
+    let byte_size = <EdwardsAffine as Zero>::zero().serialized_size();
```
```suggestion
    let byte_size = EdwardsAffine::zero().serialized_size();
```

Pratyush commented on algebra/src/tests/curves.rs:
```diff
@@ -332,40 +338,45 @@ pub fn sw_curve_serialization_test<P: SWModelParameters>(buf_size: usize) {
-            });
+            let mut serialized = vec![0; buf_size - 1];
+            let mut cursor = Cursor::new(&mut serialized[..]);
+            a.serialize(&mut cursor).unwrap_err();
```
Should there be an `assert!` here?

paberr replied:
It is implicit in the `unwrap_err()` call. Alternatively, I can replace it by `assert!([...].is_err())`.

Pratyush reviewed (commented):
Left some more nits and a couple of questions.

paberr committed:
Improve serialization size (do not require full u64 limbs to be serialised)
Minor style improvements and fixes

paberr committed:
Change Flags to enum types

paberr committed:
Split up serialization with and without flags
Implement simple derive macro
Derive serializers for parameters, keys, and proofs

paberr commented:
Okay, I changed the flags to enums and split up the serialization with and without flags into separate traits. That makes implementing serialization for keys, parameters, and proofs easier.

I also added a simple derive macro for the (de-)serialization.
Vectors are serialized by a u64 length field followed by its elements. This might not be the most optimal encoding for all cases. However, I wanted to keep it simple as `zexe` is not a serialization library. ;)
Rust requires the macro to be in a separate crate. So, location and naming thereof is up for discussion.
Currently, it is called `algebra-core-derive` and lives below `algebra-core`.

Pratyush commented on algebra/src/bls12_377/fields/tests.rs:
```diff
@@ -44,11 +46,29 @@ fn test_fq() {
+        field_serialization_test::<SW6Fq>(byte_size);
+    }
+}
+
```
Can you move this to `algebra/src/sw6/fields/tests?

Pratyush commented on groth16/src/lib.rs:
```diff
@@ -48,7 +51,7 @@ mod test;
 
 /// A proof in the Groth16 SNARK.
-#[derive(Clone)]
+#[derive(Clone, CanonicalSerialize, CanonicalDeserialize)]
```
For these impls to work, you would probably need to either make `PairingEngine: CanonicalSerialize + CanonicalDeserialize`, or you would need to adjust the derive macros to take in a separate `bounds` parameter like in [`derivative`](https://mcarton.github.io/rust-derivative/Debug.html#custom-bound) or in [`serde`](https://serde.rs/attr-bound.html).

Pratyush replied:
(The latter is preferable, but is also more work, so it's up to you)

paberr replied:
No, I don't think you need this. The only requirement for the derive macro to work is that all fields within the struct implement CanonicalSerialize/CanonicalDeserialize (and it will fail during compilation if this requirement is not given).

`E::G1Affine` and `E::G2Affine` both have the trait bound `AffineCurve`, which requires `CanonicalSerialize`/`CanonicalDeserialize` to be implemented. Thus, the derivation should work.

For reference, here's the code the compiler currently produces for exactly this derivation here:
```
impl <E: PairingEngine> CanonicalSerialize for Proof<E> {
    #[allow(unused_mut, unused_variables)]
    fn serialize<W: ::algebra_core::io::Write>(&self, writer: &mut W)
     -> Result<(), ::algebra_core::SerializationError> {
        CanonicalSerialize::serialize(&self.a, writer)?;
        CanonicalSerialize::serialize(&self.b, writer)?;
        CanonicalSerialize::serialize(&self.c, writer)?;
        Ok(())
    }
    #[allow(unused_mut, unused_variables)]
    fn serialized_size(&self) -> usize {
        let mut size = 0;
        size += CanonicalSerialize::serialized_size(&self.a);
        size += CanonicalSerialize::serialized_size(&self.b);
        size += CanonicalSerialize::serialized_size(&self.c);
        size
    }
}
impl <E: PairingEngine> CanonicalDeserialize for Proof<E> {
    #[allow(unused_mut, unused_variables)]
    fn deserialize<R: ::algebra_core::io::Read>(reader: &mut R)
     -> Result<Self, ::algebra_core::SerializationError> {
        {
            Ok(Proof{a:
                         ::algebra_core::CanonicalDeserialize::deserialize(reader)?,
                     b:
                         ::algebra_core::CanonicalDeserialize::deserialize(reader)?,
                     c:
                         ::algebra_core::CanonicalDeserialize::deserialize(reader)?,})
        }
    }
}
```

Pratyush replied:
Interesting, the default derive macros from the `std` library add trait bounds to each type parameter.

Pratyush reviewed (commented):
Thanks for your patience and the awesome work! Just left a couple of small notes.

paberr committed:
Move test to better location
